unison-cli/src/Unison/Share/Sync.hs

```diff
@@ -336,78 +335,72 @@ pull ::
   Sqlite.Connection ->
   -- | The repo+path to pull from.
   Share.Path ->
-  -- | Callback that is given the total number of entities downloaded.
-  (Int -> IO ()) ->
+  -- | Callback that is given the total number of entities downloaded, and the number of outstanding entities to
+  -- download.
+  (Int -> Int -> IO ()) ->
   IO (Either (SyncError PullError) CausalHash)
-pull httpClient unisonShareUrl conn repoPath@(Share.pathRepoName -> repoName) downloadCountCallback = catchSyncErrors do
+pull httpClient unisonShareUrl conn repoPath downloadProgressCallback = catchSyncErrors do
   getCausalHashByPath httpClient unisonShareUrl repoPath >>= \case
     Left err -> pure (Left (PullErrorGetCausalHashByPath err))
     -- There's nothing at the remote path, so there's no causal to pull.
     Right Nothing -> pure (Left (PullErrorNoHistoryAtPath repoPath))
     Right (Just hashJwt) -> do
       let hash = Share.hashJWTHash hashJwt
-      doDownload <- makeDoDownload httpClient unisonShareUrl repoName downloadCountCallback
-      tempEntities <-
+      (maybeTempEntities, downloadedSoFar) <-
         Sqlite.runTransaction conn (Q.entityLocation hash) >>= \case
-          Just Q.EntityInMainStorage -> pure Nothing
-          Just Q.EntityInTempStorage -> pure (Just (NESet.singleton hash))
-          Nothing -> downloadEntities doDownload conn (NESet.singleton hashJwt)
-      whenJust tempEntities (completeTempEntities doDownload conn)
+          Just Q.EntityInMainStorage -> pure (Nothing, 0)
+          Just Q.EntityInTempStorage -> pure (Just (NESet.singleton hash), 0)
+          Nothing -> do
+            tempEntities <- doDownload (NESet.singleton hashJwt)
+            pure (tempEntities, 1)
+      downloadedTotal <-
+        case maybeTempEntities of
+          Nothing -> pure downloadedSoFar
+          Just tempEntities -> do
+            completeTempEntities
+              doDownload
+              conn
+              (\downloaded enqueued -> downloadProgressCallback (downloaded + downloadedSoFar) enqueued)
+              tempEntities
+      downloadProgressCallback downloadedTotal 0
```

Comment:
Can we avoid displaying a progress message at all if we don't need to download anything?

I'm wondering if we should actually special-case the message when you already had everything downloaded; otherwise users might wonder if the pull successfully updated their branch when they see `Downloaded 0/0 entities`